### PR TITLE
Models and associations created

### PR DESCRIPTION
**Pull request by Yosefgeda**
- Created model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
- Set up associations between models.
- Created custom methods for:
1. Users
- A method that returns the 3 most recent posts for a given user.
2. Posts
- A method that updates the posts counter for a user.
- A method that returns the 5 most recent comments for a given post.
3. Comments
- A method that updates the comments counter for a post.
4. Likes
- A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class User < ApplicationRecord
+  has_many :comments
+  has_many :likes
+  has_many :posts
```
Comment (danmainah) [inline app/models/user.rb]: - [ ] @Yosefgeda great work adding the has_many relationship, however, you forgot to include the foreign_key referencing the user eg `has_many :posts, foreign_key: 'author_id'`, kindly consider adding foreign keys to make the association work.

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class Post < ApplicationRecord
+  belongs_to :user
```
Comment (danmainah) [inline app/models/post.rb]: - [ ] @Yosefgeda great work adding associations in the post model, however, you reference the user wrongly, instead you should use the class and author eg ``belongs_to :author, class_name: 'User'``

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Comment < ApplicationRecord
+  belongs_to :user
```
Comment (danmainah) [inline app/models/comment.rb]: - [ ]  @Yosefgeda great work adding associations in the comment model, however, you reference the user wrongly, instead, you should use the class and author eg `belongs_to :author, class_name: 'User'`

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
Comment (danmainah) [inline app/models/like.rb]: - [ ]  @Yosefgeda great work adding associations in the like model, however, you reference the user wrongly, instead, you should use the class and author eg `belongs_to :author, class_name: 'User'`

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+  def three_recent_posts
+    posts.limit(3)
```
Comment (danmainah) [inline app/models/user.rb]: - [ ] @Yosefgeda great work adding a method to return the most recent posts, however your method doesn't guarantee you will get the 3 recent post, kindly consider changing it , take a look at my suggestion below
```suggestion
    `posts.order('created_at Desc').limit(3)`
```

File: app/models/post.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  has_many :comments
+  has_many :likes
+
+  after_save :update_posts_counter
+
+  def five_recent_comments
+    comments.limit(3)
```
Comment (danmainah) [inline app/models/post.rb]: - [ ] @Yosefgeda great work adding a method for returning 5 recent comments, however your method current returns 3 comments and doenst guarantee they will be most recent, kindly check on my suggest on the `three_recent_posts` method on user model and apply the same here.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Yosefgeda  , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ Linter setup correctly and all checks are passing
✔️  Added post comment and likes counter correctly

# Required Changes ♻️

- [ ] @Yosefgeda great work on your documentation, however kindly ensure you have the correct information on how to run the rails app, the current description is not correct. You could also indicate they need to have installed Ruby and Ruby on Rails in order to run the app.
`hint: you could include bundle install to install dependencies and rails s to run the app

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class User < ApplicationRecord
+  has_many :comments
+  has_many :likes
+  has_many :posts, foreign_key: 'author_id'
```
Comment (Ghiftee) [inline app/models/user.rb]: - [ ] Good job setting the foreign key for posts as the previous reviewer requested, however, you would also need to set it for likes and comments too. e.g:
``` 
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
has_many :posts, foreign_key: 'author_id'

**Review by Ghiftee — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️♻️
Hi there @Yosefgeda 👋🏽 , 

Great work on making the changes requested by a previous reviewer 👏 
To highlight 👍🏽 
✔️ References are fixed
✔️ README is updated
✔️ Recent posts and comments are updated

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Required Changes ♻️
_Check the comments under the review 👇🏽 _

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by jerryowusu — APPROVED** [review]
Hi @Yosefgeda 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

chelsea-mason-approve


## Highlights
No linter issues :+1: 
Great effort in making required changes :+1: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**